### PR TITLE
Complexbin

### PR DESCRIPTION
**Pull request by sachinagada**

### CODE REVIEW

File: panels/panels.go [hunk 1]
```diff
@@ -212,6 +212,13 @@ func (p *Panel) help(binaryName string, args []string) error {
 	// TODO: recursive
 }
 
+func Must(sch *SubCmdHandle, err error) *SubCmdHandle {
+	if err != nil {
+		panic(fmt.Errorf("error registering subcommands: %w", err))
```
Comment (dfinkel) [inline panels/panels.go:217]: nit: `subcommand` (since we don't have a batch-registration interface)

File: panels/examples/complexbin/animal.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/vimeo/dials/panels"
+)
+
+var _ = panels.Must(p.Register("animals", defaultAnimals()))
```
Comment (dfinkel) [inline panels/examples/complexbin/animal.go:11]: nit: let's move this registration to the bottom of the file.
... this way we're not referencing some thing that hasn't been created yet.

File: panels/examples/complexbin/animal.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+func (a *animals) LongUsage(scPath []string) string {
+	return strings.Join(scPath, " ") + "\n\tget all the animals"
+}
+
+func (a *animals) DefaultConfig() interface{}      { return &animalTraits{} }
+func (a *animals) SetupParams() panels.SetupParams { return panels.SetupParams{} }
+
+func (a *animals) Run(ctx context.Context, s *panels.Something) error {
+	var sb strings.Builder
```
Comment (dfinkel) [inline panels/examples/complexbin/animal.go:43]: I don't think there's any benefit from using `strings.Builder` in this function.
We might as well write directly to `s.W`.

File: panels/examples/complexbin/animal.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		if c.Type != "" && c.Type != animal.Type {
+			continue
+		}
+
+		fmt.Fprintf(&sb, "Name: %s\tType:%s\tLegs: %d\n", n, animal.Type, animal.Legs)
+	}
+	if len(sb.String()) == 0 {
+		fmt.Fprint(&sb, "No animals found matching the specified requirements\n")
+	}
+	fmt.Fprint(s.W, sb.String())
```
Comment (dfinkel) [inline panels/examples/complexbin/animal.go:71]: Hmm, I don't think there's any benefit to using `fmt.Fprint` for a string like this over directly calling `s.W.Write([]byte())`. (or using `io.WriteString()` to do the cast for us)

**Review by dfinkel — COMMENTED** [review]
a few thoughts

**Review by dfinkel — APPROVED** [review]

File: panels/examples/complexbin/animal.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		if c.Legs != 0 && c.Legs != animal.Legs {
+			continue
+		}
+
+		if c.Type != "" && c.Type != animal.Type {
+			continue
+		}
+
+		foundMatch = true
+		s.W.Write([]byte(fmt.Sprintf("Name: %s\tType:%s\tLegs: %d\n", n, animal.Type, animal.Legs)))
```
Comment (dfinkel) [inline panels/examples/complexbin/animal.go:65]: Oh, I missed this!
The `fmt.Fprintf` calls were fine, it's just that writing to a string-builder to then immediately write out its contents in an example like this was unnecessary complexity.
Let's switch this to:
```suggestion
		fmt.Fprintf(s.W,"Name: %s\tType:%s\tLegs: %d\n", n, animal.Type, animal.Legs)
```

Comment (sachinagada) [reply panels/examples/complexbin/animal.go]: Good catch!

**Review by dfinkel — APPROVED** [review]

**Pull request merged**